jax_md/energy.py

```diff
@@ -891,6 +893,285 @@ def compute_fn(R, neighbor, **kwargs):
+  raise NotImplementedError('The model without neighbor lists '
+                            ' is not implemented yet.')
+
+def tersoff_neighbor_list(displacement: DisplacementFn,
+  box_size: float,
+  params: Array,
+  species: Optional[Array]=None,
+  dr_threshold: float=0.5,
+  frac_coords: bool=True,
+  format: NeighborListFormat=partition.Dense,
+  **neighbor_kwargs
+) -> Tuple[NeighborFn, Callable[[Array, NeighborList], Array]]:
+  """Computes the Tersoff potential.
+
+  The Tersoff potential [1] which is commonly used to model 
+  semiconducting materials. The Tersoff potential was originally proposed to
+  model various types of lattice with a simple functional form.
+  For this reason, Tersoff model was introduced bond-order function
+  to determine the strength of repulsive and attractive forces between atoms.
+
+  Args:
+    displacement: The displacement function for the space.
+
+  Returns:
+    A function that computes the total energy.
+
+  [1] J. Tersoff "New empirical approach for the structure and energy of
+  covalent systems" Physical review B 37.12 (1988): 6991.
+  """
+  # check number of parameters set
+  if species is None:
+    params = params[0]
+    nparams = 1
+  else:
+    raise NotImplementedError('Multiple species were not implemented yet.')
+
+  # define a repulsive and an attractive function with given parameters
+  repulsive_fn = partial(_ters_repulsive, 
+                         params['A'], params['lam1'], 
+                         params['R'], params['D'])
+  attractive_fn = partial(_ters_attractive, 
+                          params['B'], params['lam2'], 
+                          params['R'], params['D'],
+                          params['cTf'], params['dTf'], params['hTf'], 
+                          params['lam3'], params['beta'],
+                          params['nTf'], params['mTf'])
+
+  # define a neighbor function.
+  # todo: other neighbor list construction method will be implemented.
+  if format is partition.Dense:
+    neighbor_fn = partition.neighbor_list(displacement,
+                                          box_size,
+                                          params['R']+params['D'],
+                                          dr_threshold,
+                                          disable_cell_list=True,
+                                          fractional_coordinates=frac_coords,
+                                          format=format,
+                                          **neighbor_kwargs)
+  else:
+    raise NotImplementedError('Tersoff potential only implemented '
+                                'with Dense neighbor lists.')
+
+  # define compute functions
+  # todo: is neighboring mask necessary? 
+  def compute_fn(R, neighbor, **kwargs):
+    d = partial(displacement, **kwargs)
+    mask = partition.neighbor_list_mask(neighbor, mask_self=True)
+    mask_ijk = mask[:, None, :] * mask[:, :, None]
+
+    if neighbor.format is partition.Dense:
```

Comment:
You already check to make sure the neighbor list format is dense above.

Reply:
Thanks for your comment. I deleted that part.